### ISSUE TITLE
vim.ui enhancements: proposal and tracking issue

### ISSUE THREAD

**Issue by gpanders** [neovim/neovim#25514]
This ticket specifically tracks enhancements and modifications to `vim.ui` proposed in #25272 and related issues.

# Problem

Creating floating windows with all of the visual goodies that most users expect and want requires a lot of boilerplate. This boilerplate has coalesced into `vim.lsp.util.open_floating_preview`, but such functionality should be both (1) pruned down and curated into a more thoughtful and coherent interface, and (2) moved out of `vim.lsp.util` and into `vim.ui`.

# Solution

Create `vim.ui.open_win` that handles the use case of creating windows of all kinds in an ergonomic way.

## Requirements

* `vim.ui.open_win` should be able to open floating windows as well as splits (horizontal and vertical) and tabs.
* `vim.ui.open_win` should substantially reduce boilerplate for creating and opening floating windows.

## Proposed interface

```lua
--- @param contents string[]|int An array of strings containing the lines
---                              of text to display in the window OR an
---                              integer representing a buffer number to
---                              display in the window
--- @param opts table Options table (see details below)
--- @return number |winid| of new window
vim.ui.open_win(contents, {
  -- Opts table is a "sum type" / "union", the type of window it
  -- creates is based on the top-level key provided. It as an error
  -- to provide more than one top-level key
  float = {
    -- Create a floating window. All keys from |nvim_open_win()| are valid in this table.
    -- Omitted keys use default values from |vim.ui.config()|.
    height = 100,
    width = 50,
  },

  split = {
    -- Create a split
    vertical = true | false, -- defaults to false
  },

  tab = {
    -- Create a new tab
  }
})
```

In addition, create `vim.ui.config()` that sets default values for some options. This should work similarly to `vim.diagnostic.config()`. Plugin authors can use `vim.ui.open_win()` and omit values to use the defaults supplied by the user in `vim.ui.config()`:

```lua
vim.ui.config({
  float = {
    border = "rounded",
  },
  split = {
    -- Prefer vertical splits by default
    vertical = true,
  },
})
```

`vim.ui.config()` is not limited to configuring options for `vim.ui.open_win()`, but covers any user-definable default/preference for anything under `vim.ui`.

**Comment by mfussenegger** [comment]
> > Any reason for a vim.ui.config table instead of going with new regular vim options?
> 
> Because we try to keep things in Lua if possible (and do you really want to touch Vim options? especially window options? especially global-local window options?)
> 
> I believe the question of Lua config table vs. Vim option is a more general (and much thornier) issue and should be handled separately (and apply to `vim.diagnostic` config as well).


This is exactly why I bring it up:

- I think we should have this more general discussion :)
- diagnostic configuration has global and "namespace" options (opposed to buffer options or window options). 1 diagnostic provider=1 namespace. Often a language server or linter. The options for a floating window would fit more into the existing window/buffer option model. Because a floating window is a window and it displays a buffer.

**Comment by clason** [comment]
So would these be separate window option, or one giant dict option?

**Comment by justinmk** [comment]
> Because we try to keep things in Lua if possible

That is not a guideline nor a goal.

>  (and do you really want to touch Vim options? especially window options? especially global-local window options?)

We are investing in the options subsystem: https://github.com/neovim/neovim/pull/25394 
If options make sense as editor options then "because Lua" can't be the reason for avoiding the options subsystem.

`vim.foo.config` (`vim.diagnostic.config`) is a convention for cases that don't make sense as "editor options" (at least until we gain dict options and perhaps eventually namespace options).

**Comment by justinmk** [comment]
From https://github.com/neovim/neovim/issues/25272 : 

> > Should the open_float features be owned by nvim_open_win instead of adding vim.ui.open_float ?
> 
> I've no objection, though having more of this done in Lua rather than in C might be nice.

API logic can be implemented in Lua (and we should try to make that more ergonomic). [Example](https://github.com/neovim/neovim/blob/29fe883aa9166bdbcae3f935523c75a8aa56fe45/runtime/lua/vim/_editor.lua#L169).

**Comment by gpanders** [comment]
> What's the motivation for putting split/tab handling in the same API? (What's wrong with `vim.cmd.split()`?)

Ergonomics. There's nothing "wrong" with `vim.cmd.split`, just as there's nothing "wrong" with `vim.api.nvim_open_win`.

Both floating windows and splits have a similar series of steps:

1. create a scratch buffer
2. create the split window. This is the only part that really differs between floats and splits (`nvim_open_win` vs `vim.cmd.split`)
3. retrieve the window ID (you get this is as the return value of `nvim_open_win`, but requires a separate API call with a split)
4. set the window's buffer to the scratch buffer

If we provide some "sugar" around creating floats (which we already do with `open_floating_preview`) that abstracts most of this boilerplate, why would we not also do the same for other kinds of windows?

> +1 for having a config for some defaults, would be nice if plugins didn't need to expose border settings. Any reason for a vim.ui.config table instead of going with new regular vim options?

No, I don't particularly care either way. If regular Vim options can support the level of flexibility we need then I'd actually prefer that.

**Comment by justinmk** [comment]
> Creating both floating windows and splits require a similar series of steps:

Ok with those steps outlined, I can see how this wouldn't fit into `nvim_open_win`.

**Comment by mfussenegger** [comment]
> Ergonomics. There's nothing "wrong" with vim.cmd.split, just as there's nothing "wrong" with vim.api.nvim_open_win.

Fair point, but also slightly derailing from the given problem statement.
Looks like there are two separate points:

- See what additional options `nvim_open_win` could support to get rid of the floating window positioning logic in `open_floating_preview` and `make_floating_popup_options`. This is kinda independent from the proposed `open_win` but would compose (due to the `float = {}` opts)

- The ergonomics for general window creation. I think the uniqueness handling that `open_floating_preview` currently takes car of could also go into this. E.g. `open_win` could create an object where you can call `:toggle()`, `:open()`, `:close()` on. It would move some of the logic for the uniqueness handling to the consumers, but I think this separation of concerns makes sense, after all the consumer needs to decide _which_ window should have what kind of uniqueness semantics.


> retrieve the window ID (you get this is as the return value of nvim_open_win, but requires a separate API call with a split)
> set the window's buffer to the scratch buffer

Just to complete the picture: `nvim_open_win` takes the buffer as argument. For floating window it is currently just step 1 (create buffer) and step 2 (call `nvim_open_win(buf, ...)`)

Most of the logic around opening floating windows seems to be around calculating the position, and configuring the buffer.

**Comment by justinmk** [comment]
> E.g. `open_win` could create an object where you can call `:toggle()`, `:open()`, `:close()` on.

Let's avoid OOP for win/buf/tab objects (at least), it raises many questions for little gain. "Toggle" could apply to any window (including non-floats) so we could add `nvim_win_toggle`. It's trivial to fake OOP-style interface with [UFCS](https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax) which is what API clients already do.

**Comment by mfussenegger** [comment]
https://github.com/neovim/neovim/pull/25550 addresses the split window part

**Comment by willothy** [comment]
>  "Toggle" could apply to any window (including non-floats) so we could add `nvim_win_toggle`

+1 to `nvim_win_toggle`, I'd be interested in working on that. I do think it could get tricky with edge cases though, as we'd probably have to handle windows being toggled once (closed) but never reopened to avoid leaking memory when saving window states. There's also the case where there's a split window that is closed by toggle, then its previous parent frame is removed by closing its "sibling" windows. Working out where to re-insert the window into the layout would be tough. Sorry if that's outside the scope of this issue, just my initial thoughts on the idea :)

**Comment by justinmk** [comment]
> There's also the case where there's a split window that is closed by toggle, then its previous parent frame is removed by closing its "sibling" windows.

Could avoid that by keeping the window but set its width/height to zero. Of course that means other things need to be handled, e.g. `CTRL-w-...` should skip over such windows.

**Comment by willothy** [comment]
> Could avoid that by keeping the window but set its width/height to zero. 

True! I think that takes care of both cases. Will look into this after `nvim_open_win` updates are done :)

**Comment by brianhuster** [comment]
Just my 2 cents, but I think it would be nice to have a `floating` option for `vim.ui.select()`, similar to Vim 8's `popup_menu()`

**Comment by justinmk** [comment]
Renamed this, to focus on the "open a window/popup easily" topic. The discussion already seems focused on that.
